packages/synapse-react-client/src/components/WikiMarkdownEditor/WikiMarkdownEditor.tsx

```diff
@@ -0,0 +1,256 @@
+import { Alert, Box, Button, TextField } from '@mui/material'
+import { ObjectType, WikiPageKey } from '@sage-bionetworks/synapse-types'
+import { noop } from 'lodash-es'
+import React, { useEffect, useMemo, useState } from 'react'
+import {
+  CreateWikiPageInput,
+  UpdateWikiPageInput,
+  useCreateWikiPage,
+  useGetRootWikiPageKey,
+  useGetWikiPage,
+  useUpdateWikiPage,
+} from '../../synapse-queries'
+import ConfirmationDialog from '../ConfirmationDialog'
+import { DialogBase } from '../DialogBase'
+import { SynapseSpinner } from '../LoadingScreen/LoadingScreen'
+import { MarkdownEditor } from '../Markdown'
+
+export const UNSAVED_CHANGES = 'Unsaved Changes'
+export const NAVIGATE_AWAY_CONFIRMATION_MESSAGE =
+  'Any unsaved changes may be lost. Are you sure that you would like to navigate away from this editor?'
+export const ERROR_LOADING_WIKI_FAILED = 'Failed to load the wiki page: '
+export const ERROR_SAVING_WIKI =
+  'Could not save your changes. It is recommended that you copy your version of the wiki text so that it is not lost. '
+
+export type WikiMarkdownEditorProps = {
+  ownerObjectType: ObjectType
+  ownerObjectId: string
+  // if wikiPageId is undefined, will get (or create) the root WikiPage for this ownerObject
+  // otherwise, will get the WikiPage with the specified wikiPageId
+  wikiPageId?: string
+  onCancel?: () => void
+  onSave?: () => void
+  // TODO: SWC-6774 - expose delete prop
+  // showDeleteButton: boolean
+}
+
+export const WikiMarkdownEditor: React.FunctionComponent<
+  WikiMarkdownEditorProps
+> = (props: WikiMarkdownEditorProps) => {
+  // TODO: SWC-6774 - get showDeleteButton from props
+  const showDeleteButton = false
+  const {
+    ownerObjectType,
+    ownerObjectId,
+    wikiPageId: initialWikiPageId,
+    onCancel = noop,
+    onSave = noop,
+  } = props
+
+  const [wikiPageId, setWikiPageId] = useState<string | undefined>(
+    initialWikiPageId,
+  )
+  const [open, setOpen] = useState<boolean>(true)
+  const [title, setTitle] = useState<string>('')
+  const [markdown, setMarkdown] = useState<string>('')
+  const [showConfirmCancelDialog, setShowConfirmDialog] =
+    useState<boolean>(false)
+  const [error, setError] = useState<string | null>(null)
+
+  const { data: rootWikiPageKey, error: errorLoadingRootWikiPageKey } =
+    useGetRootWikiPageKey(ownerObjectType, ownerObjectId, {
+      enabled: wikiPageId === undefined,
+    })
+
+  const { mutate: createWikiPage } = useCreateWikiPage({
+    onSuccess: wikiPage => {
+      setError(null)
+      setWikiPageId(wikiPage.id)
+    },
+    onError: error => {
+      setError(error.reason)
+    },
+  })
+
+  useEffect(() => {
+    setError(null)
+
+    // an existing root WikiPageKey was found
+    if (rootWikiPageKey) {
+      setWikiPageId(rootWikiPageKey.wikiPageId)
+    }
+
+    // root WikiPageKey was not found (so we have create permission),
+    // then create a root WikiPage for this ownerObject
+    if (rootWikiPageKey === null) {
```

Comment:
And attempt to create a new root WikiPage, if one does not already exist

Reply:
Why do we unconditionally create this in an effect instead of waiting for the user to try to explicitly save a wiki? Is the idea that, if this is open, you already chosen to create the wiki page?

Reply:
_If_ this effect is the right place to create the wiki (which I am not convinced), how are we assured that this hook doesn't fire twice, triggering two 'create' calls? In that case I think we should also prevent calling `createWikiPage` if the existing `createWikiPage` mutation `isPending`

Reply:
I was following the SWC implementation, which [creates a root WikiPage](https://github.com/Sage-Bionetworks/SynapseWebClient/blob/9085d9d2f9644ba7ceb9db9624b3a48e52a269a4/src/main/java/org/sagebionetworks/web/client/widget/entity/WikiMarkdownEditor.java#L132) if one did not exist when the editor was opened. I was also surprised by that flow initially, so I'm happy to wait to create the wiki until the user explicitly tries to save the wiki instead!

Good point about triggering two 'create' calls -- if we stick with the effect, I'll add the guard against a pending `createWikiPage` call!